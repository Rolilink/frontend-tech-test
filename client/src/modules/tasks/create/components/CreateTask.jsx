import React from 'react';
import PropTypes from 'prop-types';
import LoadingMessage from '../../../common/LoadingMessage';
import ErrorMessage from '../../../common/ErrorMessage';
import TaskForm from '../../../common/TaskForm';

const getInitialState = () => ({
  task: {
    title: '',
    description: '',
  },
});

export default class CreateTask extends React.Component {
  static propTypes = {
    createTask: PropTypes.func.isRequired,
    changeMode: PropTypes.func.isRequired,
    className: PropTypes.string,
    isFetching: PropTypes.bool,
    error: PropTypes.any,
  }

  static defaultProps = {
    className: '',
    error: null,
    isFetching: false,
  }

  constructor(props) {
    super(props);
    this.state = getInitialState();
  }

  get taskForm() {
    const { className } = this.props;
    const { task } = this.state;

    return (
      <TaskForm
        className={`CreateTask-TaskForm ${className}`}
        onTitleUpdate={title => this.setState({ task: { title } })}
        onDescriptionUpdate={description => this.setState({ task: { description } })}
        onSubmit={() => this.submitTask()}
        onCancel={() => this.clearMode()}
        task={task}
      />
    );
  }

  getErrorMessage(error) {
    return (
      <div>{error}</div>
    );
  }

  getLoadingMessage() {
    return (
      <div>Creating Task</div>
    );
  }

  submitTask() {
    const { task } = this.state;

    this.props.createTask(task);
  }

  clearMode() {
    this.setState(getInitialState());
    this.props.changeMode(null);
  }

  render() {
    const { isFetching, error } = this.props;

    return (
      <LoadingMessage
        isLoading={isFetching}
        renderMessage={this.getLoadingMessage}
      >
        <ErrorMessage
          error={error}
          renderMessage={this.getErrorMessage}
        >
          {this.taskForm}
        </ErrorMessage>
      </LoadingMessage>
    );
  }
}
