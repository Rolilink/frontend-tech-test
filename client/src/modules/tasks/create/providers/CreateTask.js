import { connect } from 'react-redux';
import { createTask, selectors } from '../../redux/';

const { getCreateTaskProps } = selectors;

function mapStateToProps(state) {
  return getCreateTaskProps(state);
}

const mapDispatchToProps = {
  createTask,
};

export default connect(mapStateToProps, mapDispatchToProps);
