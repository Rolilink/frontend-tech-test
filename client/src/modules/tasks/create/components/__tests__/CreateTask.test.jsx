import React from 'react';
import { shallow, mount } from 'enzyme';
import serializer from 'enzyme-to-json/serializer';

import CreateTask from '../CreateTask';

expect.addSnapshotSerializer(serializer);

describe('Tasks.Create.CreateTask', () => {
  let createTask;
  let changeMode;
  let isFetching;
  let error;
  let defaultState;

  beforeEach(() => {
    createTask = jest.fn();
    changeMode = jest.fn();
    isFetching = false;
    error = null;
    defaultState = {
      task: {
        description: '',
        title: '',
      },
    };
  });

  it('should render a form', () => {
    const wrapper = mount(
      <CreateTask
        {...{
          createTask,
          changeMode,
        }}
      />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should render an error message when error prop is present', () => {
    error = 'error';

    const wrapper = mount(
      <CreateTask
        {...{
          createTask,
          changeMode,
          isFetching,
          error,
        }}
      />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should render an LoadingMessage when isFetching prop is present', () => {
    isFetching = true;

    const wrapper = mount(
      <CreateTask
        {...{
          createTask,
          changeMode,
          isFetching,
          error,
        }}
      />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should return an error message when calling #getErrorMessage', () => {
    const wrapper = shallow(
      <CreateTask
        {...{
          createTask,
          changeMode,
          isFetching,
          error,
        }}
      />,
    );

    expect(wrapper.instance().getErrorMessage('message')).toMatchSnapshot();
  });

  it('should return a loading message when calling #getLoadingMessage', () => {
    const wrapper = shallow(
      <CreateTask
        {...{
          createTask,
          changeMode,
          isFetching,
          error,
        }}
      />,
    );

    expect(wrapper.instance().getLoadingMessage()).toMatchSnapshot();
  });

  it('should call props.changeMode with null when calling #clearMode', () => {
    const wrapper = shallow(
      <CreateTask
        {...{
          createTask,
          changeMode,
          isFetching,
          error,
        }}
      />,
    );

    wrapper.instance().clearMode();

    expect(changeMode).toHaveBeenCalledWith(null);
  });

  it('should set the state to its default when calling #clearMode', () => {
    const task = {
      title: 'test',
      description: 'test description',
    };

    const wrapper = shallow(
      <CreateTask
        {...{
          createTask,
          changeMode,
          isFetching,
          error,
        }}
      />,
    );

    wrapper.instance().setState({ task });
    wrapper.instance().clearMode();

    expect(wrapper.instance().state).toEqual(defaultState);
  });

  it('should update the title when calling #updateTitle', () => {
    const updatedTitle = 'updated title';

    const wrapper = shallow(
      <CreateTask
        {...{
          createTask,
          changeMode,
          isFetching,
          error,
        }}
      />,
    );

    wrapper.instance().updateTitle(updatedTitle);

    expect(wrapper.instance().state.task.title).toBe(updatedTitle);
  });

  it('should update the description state when calling #updatedDescription', () => {
    const updatedDescription = 'updated description';

    const wrapper = shallow(
      <CreateTask
        {...{
          createTask,
          changeMode,
          isFetching,
          error,
        }}
      />,
    );

    wrapper.instance().updateDescription(updatedDescription);

    expect(wrapper.instance().state.task.description).toBe(updatedDescription);
  });

  it('should call props.createTask with task when calling #submitTask', () => {
    createTask = jest.fn(() => (Promise.resolve()));
    const task = {
      title: 'test',
      description: 'test description',
    };

    const wrapper = shallow(
      <CreateTask
        {...{
          createTask,
          changeMode,
          isFetching,
          error,
        }}
      />,
    );

    wrapper.instance().setState({ task });
    wrapper.instance().submitTask();

    expect(createTask).toHaveBeenCalledWith(task);
  });

  it('should set the default state wwhen calling #submitTask', () => {
    createTask = jest.fn(() => (Promise.resolve()));
    const task = {
      title: 'test',
      description: 'test description',
    };

    const wrapper = shallow(
      <CreateTask
        {...{
          createTask,
          changeMode,
          isFetching,
          error,
        }}
      />,
    );

    wrapper.instance().setState({ task });
    return wrapper.instance().submitTask().then(() => {
      expect(wrapper.instance().state).toEqual(defaultState);
    });
  });
});
