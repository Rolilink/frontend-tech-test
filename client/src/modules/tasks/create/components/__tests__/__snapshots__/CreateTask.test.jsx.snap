// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks.Create.CreateTask should render a form 1`] = `
<CreateTask
  changeMode={[Function]}
  className=""
  createTask={[Function]}
  error={null}
  isFetching={false}
>
  <LoadingMessage
    className=""
    isLoading={false}
    renderMessage={[Function]}
  >
    <ErrorMessage
      className=""
      error={null}
      renderMessage={[Function]}
    >
      <TaskForm
        className="CreateTask-TaskForm "
        onCancel={[Function]}
        onDescriptionChange={[Function]}
        onSubmit={[Function]}
        onTitleChange={[Function]}
        task={
          Object {
            "description": "",
            "title": "",
          }
        }
      >
        <Form
          bsClass="form"
          className="TasksForm CreateTask-TaskForm "
          componentClass="form"
          horizontal={false}
          inline={false}
          onSubmit={[Function]}
        >
          <form
            className="TasksForm CreateTask-TaskForm  "
            onSubmit={[Function]}
          >
            <FormGroup
              bsClass="form-group"
              className="TasksForm-Title"
            >
              <div
                className="TasksForm-Title form-group"
              >
                <ControlLabel
                  bsClass="control-label"
                  srOnly={false}
                >
                  <label
                    className="control-label"
                  >
                    Title:
                  </label>
                </ControlLabel>
                <FormControl
                  bsClass="form-control"
                  className="TasksForm-TitleInput"
                  componentClass="input"
                  onChange={[Function]}
                  type="text"
                  value=""
                >
                  <input
                    className="TasksForm-TitleInput form-control"
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                </FormControl>
              </div>
            </FormGroup>
            <FormGroup
              bsClass="form-group"
              className="TasksForm-Description"
            >
              <div
                className="TasksForm-Description form-group"
              >
                <ControlLabel
                  bsClass="control-label"
                  srOnly={false}
                >
                  <label
                    className="control-label"
                  >
                    Description:
                  </label>
                </ControlLabel>
                <FormControl
                  bsClass="form-control"
                  className="TasksForm-DescriptionInput"
                  componentClass="input"
                  onChange={[Function]}
                  type="text"
                  value=""
                >
                  <input
                    className="TasksForm-DescriptionInput form-control"
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                </FormControl>
              </div>
            </FormGroup>
            <FormGroup
              bsClass="form-group"
              className="TasksForm-Actions"
            >
              <div
                className="TasksForm-Actions form-group"
              >
                <input
                  className="btn btn-primary TasksForm-SubmitButton"
                  disabled={true}
                  type="submit"
                  value="Submit"
                />
                <Button
                  active={false}
                  block={false}
                  bsClass="btn"
                  bsStyle="danger"
                  className="btn btn-primary TasksForm-CancelButton"
                  disabled={false}
                  onClick={[Function]}
                >
                  <button
                    className="btn btn-primary TasksForm-CancelButton btn btn-danger"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                  >
                    Cancel
                  </button>
                </Button>
              </div>
            </FormGroup>
          </form>
        </Form>
      </TaskForm>
    </ErrorMessage>
  </LoadingMessage>
</CreateTask>
`;

exports[`Tasks.Create.CreateTask should render an LoadingMessage when isFetching prop is present 1`] = `
<CreateTask
  changeMode={[Function]}
  className=""
  createTask={[Function]}
  error={null}
  isFetching={true}
>
  <LoadingMessage
    className=""
    isLoading={true}
    renderMessage={[Function]}
  >
    <div
      className="LoadingMessage "
    >
      <div
        className="LoadingMessage-Spinner"
      >
        <Spinner>
          <i
            className="fa fa-spinner fa-spin"
          />
        </Spinner>
      </div>
      <div
        className="LoadingMessage-Message"
      >
        <div>
          Creating Task
        </div>
      </div>
    </div>
  </LoadingMessage>
</CreateTask>
`;

exports[`Tasks.Create.CreateTask should render an error message when error prop is present 1`] = `
<CreateTask
  changeMode={[Function]}
  className=""
  createTask={[Function]}
  error="error"
  isFetching={false}
>
  <LoadingMessage
    className=""
    isLoading={false}
    renderMessage={[Function]}
  >
    <ErrorMessage
      className=""
      error="error"
      renderMessage={[Function]}
    >
      <div
        className="ErrorMessage "
      >
        <div
          className="ErrorMessage-Sign"
        >
          Here goes an error sign
        </div>
        <div
          className="ErrorMessage-Message"
        >
          <div>
            error
          </div>
        </div>
      </div>
    </ErrorMessage>
  </LoadingMessage>
</CreateTask>
`;

exports[`Tasks.Create.CreateTask should return a loading message when calling #getLoadingMessage 1`] = `
<div>
  Creating Task
</div>
`;

exports[`Tasks.Create.CreateTask should return an error message when calling #getErrorMessage 1`] = `
<div>
  message
</div>
`;
